### PR TITLE
ODIM-6787- Load all subscription information's to cache

### PR DESCRIPTION
**Pull request by chauberahul1993**

### CODE REVIEW

File: lib-persistence-manager/persistencemgr/redis.go [hunk 1]
```diff
@@ -1722,3 +1722,111 @@ func (p *ConnPool) DeleteAggregateHosts(index, aggregateID string) error {
 	}
 	return nil
 }
+
```
Comment (bshetty) [inline lib-persistence-manager/persistencemgr/redis.go:1725]: Replace three functions GetAllDeviceSubscription, GetAllEvtSubscriptions and GetAggregateHosts by these 
// This function retrieves all data for a given index from sorted sets
// This maybe used to get all event/device subscriptions and aggregate hosts 
func (p *ConnPool) GetAllDataByIndex(index string) ([]string, error) {
	dList, ferror := getAllDataFromSortedList(index)
	if ferror != nil {
		return []string{}, ferror
	}
	EvtSubscriptions, extracterr := getDataAsStringList(dList)
	if extracterror != nil {
		return []string{}, ferror
	}

	return EvtSubscriptions, nil
}

func getAllDataFromSortedSet(index string) ( data interface{}, err error) {
	readConn := p.ReadPool.Get()
	defer readConn.Close()
	const cursor float64 = 0

	d, getErr := readConn.Do("ZCOUNT", index, 0, 0)
	if getErr != nil {
		return nil, fmt.Errorf("Unable to fetch count of data for index : " +
			index + " : " + 
			getErr.Error())
	}
	countData := d.(int64)

	data, getErr = readConn.Do("ZSCAN", index, cursor, "MATCH", 
				"*", "COUNT", countData)
	if getErr != nil {
		return []string{}, 
			fmt.Errorf("Error while fetching data for " + index + " : " + getErr.Error())
	}
	return data, nil
}

func getDataAsStringList( d interface{}) ([]string, error){ 
	var dataList []string

	if len(d.([]interface{})) > 1 {
		data, err := redis.Strings(d.([]interface{})[1], getErr)
		if err != nil {
			return []string{}, fmt.Errorf("error while marshaling data : " + err.Error())
		}
		for i := 0; i < len(data); i++ {
			if data[i] != "0" {
				dataList = append(getList, data[i])
			}
		}
	}
	return  dataList, nil;
}

File: svc-events/events/subscriptionCache.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	emptyOriginResourceToSubscriptionsMap = make(map[string]map[string]bool)
+	systemIdToAggregateIdsMap             = make(map[string]map[string]bool)
+	eventSourceToManagerIDMap             = make(map[string]string)
+	managerIDToSystemIDsMap               = make(map[string][]string)
+	managerIDToChassisIDsMap              = make(map[string][]string)
+)
+
+func LoadSubscriptionData() {
+	l.Log.Debug("Event data load initialized ")
+	t := time.Now()
```
Comment (bshetty) [inline svc-events/events/subscriptionCache.go:28]: remove timing code from everywhere. Do NOT have these in production code/committed code

File: svc-events/events/subscriptionCache.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	"strings"
+	"time"
+
+	l "github.com/ODIM-Project/ODIM/lib-utilities/logs"
+	"github.com/ODIM-Project/ODIM/svc-events/evmodel"
+	uuid "github.com/satori/go.uuid"
+)
+
+var (
+	subscriptionsCache                    = make(map[string]evmodel.Subscription)
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:15]: just declare here dont define.

File: svc-events/events/subscriptionCache.go (hunk 2 above)
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:28]: remove all timing data used for debugging

File: svc-events/events/subscriptionCache.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+func LoadSubscriptionData() {
+	l.Log.Debug("Event data load initialized ")
+	t := time.Now()
+	defer l.Log.Debug("Time take to read Complete LoadSubscriptionData ", time.Since(t))
+	getAllSubscriptions()
+	getAllAggregates()
+	getAllDeviceSubscriptions()
+}
+func getAllSubscriptions() {
+	subscriptionsCache = make(map[string]evmodel.Subscription)
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:35]: here again construct the map only after knowing the size. Else we keep extending each of these maps multiple times

File: svc-events/events/subscriptionCache.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+	for _, device := range deviceSubscriptionList {
+		devSub := strings.Split(device, "||")
+		if strings.Contains(devSub[0], "Collection") {
+			continue
+		} else {
+			updateCatchDeviceSubscriptionData(devSub[0], evmodel.GetSliceFromString(devSub[2]))
+		}
+	}
+}
+func updateCatchDeviceSubscriptionData(key string, originResources []string) {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:74]: add empty line and comment above

File: svc-events/events/subscriptionCache.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	systemId := originResources[0][strings.LastIndexByte(originResources[0], '/')+1:]
+	eventSourceToManagerIDMap[key] = systemId
+}
+
+func loadSubscriptionCacheData(sub evmodel.SubscriptionResource) {
+	if len(sub.EventDestination.OriginResources) == 0 && sub.SubscriptionID != "0" {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:80]: Don't need to fill all of these separately. Just set the two attributes in a if/else block .

File: svc-events/events/subscriptionCache.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+				OriginResources:      sub.EventDestination.OriginResources,
+				DeliveryRetryPolicy:  sub.EventDestination.DeliveryRetryPolicy,
+			}
+			addSubscriptionCache(host, subCache.Id)
+			subscriptionsCache[subCache.Id] = subCache
+		}
+	}
+}
+
+//addSubscriptionCache add subscription in corresponding cache based on key type
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:113]: Why cant we fill this in bulk and avoid make/extend of the maps??

File: svc-events/events/subscriptionCache.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+				data := make(map[string]bool)
+				data[subId] = true
+				systemToSubscriptionsMap[key] = data
+			}
+			return
+		}
+	}
+}
+
+func addEmptyOriginSubscriptionCache(subscriptionId string) {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:146]: see why we cant do this in bulk

File: svc-events/events/subscriptionCache.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	data, isExists := emptyOriginResourceToSubscriptionsMap["0"]
+	if isExists {
+		data[subscriptionId] = true
+		emptyOriginResourceToSubscriptionsMap["0"] = data
+	} else {
+		emptyOriginResourceToSubscriptionsMap["0"] = map[string]bool{subscriptionId: true}
+	}
+}
+func addAggregateSubscriptionCache(key, subId string) {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:155]: ditto

File: svc-events/events/subscriptionCache.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+	data, isExists := aggregateIdToSubscriptionsMap[key]
+	if isExists {
+		data[subId] = true
+		aggregateIdToSubscriptionsMap[key] = data
+	} else {
+		aggregateIdToSubscriptionsMap[key] = map[string]bool{subId: true}
+	}
+}
+
+func getAllAggregates() {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:165]: This code is duplicated in multiple places.... refactor

File: svc-events/events/subscriptionCache.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+		if isExists {
+			aggregateIds[aggregateUrl] = true
+			systemIdToAggregateIdsMap[ids.OdataID] = aggregateIds
+		} else {
+			systemIdToAggregateIdsMap[ids.OdataID] = map[string]bool{aggregateUrl: true}
+		}
+	}
+}
+
+//getSourceId function return system id corresponding host, if not found then return host
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:199]: explain this logic

File: svc-events/events/subscriptionCache.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+func getAllAggregates() {
+	systemIdToAggregateIdsMap = make(map[string]map[string]bool)
+
+	aggregateUrls, err := evmodel.GetAllAggregates()
+	if err != nil {
+		l.Log.Debug("Exception getting aggregates url list ", err)
+		return
+	}
+	if len(aggregateUrls) == 0 {
```
Comment (Bharath-KKB) [inline svc-events/events/subscriptionCache.go:173]: as above

File: svc-events/events/subscriptionCache.go (hunk 4 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go:35]: done

File: svc-events/events/subscriptionCache.go (hunk 5 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go:74]: done

File: svc-events/events/subscriptionCache.go (hunk 6 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go:80]: done

**Review by Bharath-KKB — COMMENTED** [review]
update as per all the comments and discuss the two files with me

**Review by sriteja-sabbineni — APPROVED** [review]

**Review by amar-shalgar — APPROVED** [review]

File: svc-events/events/subscriptionCache.go (hunk 8 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

File: svc-events/events/subscriptionCache.go (hunk 9 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

File: svc-events/events/subscriptionCache.go (hunk 10 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go:121]: done

File: svc-events/events/subscriptionCache.go (hunk 12 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

File: lib-dmtf/model/events.go [hunk 13]
```diff
@@ -239,8 +244,93 @@ const (
...
+
+	// DeliveryTerminateAfterRetries : The subscription is terminated after the
+	// maximum number of retries is reached.
+	DeliveryTerminateAfterRetries DeliveryRetryPolicy = "TerminateAfterRetries"
 )
 
+//IsValidSubscriptionType validate subscription type is valid,
+func (subscriptionType SubscriptionType) IsValidSubscriptionType() bool {
+	switch subscriptionType {
+	case SubscriptionTypeRedFishEvent, SubscriptionTypeOEM, SubscriptionTypeSNMPInform, SubscriptionTypeSNMPTrap, SubscriptionTypeSyslog, SubscriptionTySubscriptionTypeSSE:
```
[the remark this answers is not in the document]
Comment (chauberahul1993) [reply lib-dmtf/model/events.go]: done

File: lib-dmtf/model/events.go [hunk 14]
```diff
@@ -239,8 +244,93 @@ const (
...
+}
+
+func (eventType EventType) ToString() string {
+	return string(eventType)
+}
+
+//IsValidEventType return true if event type is valid
+func (eventType EventType) IsValidEventType() bool {
+	switch eventType {
+	case EventTypeAlert, EventTypeMetricReport, EventTypeOther, EventTypeResourceRemoved,
```
[the remark this answers is not in the document]
Comment (chauberahul1993) [reply lib-dmtf/model/events.go]: done

File: svc-events/events/publishevttosubscriber.go [hunk 15]
```diff
@@ -138,7 +138,7 @@ func (e *ExternalInterfaces) PublishEventsToDestination(data interface{}) bool {
 	if err != nil {
 		l.Log.Info("No Aggregate subscription Found ", err)
 	}
-	var aggregateSubscriptionList []evmodel.Subscription
+	var aggregateSubscriptionList []evmodel.SubscriptionResource
```
[the remark this answers is not in the document]
Comment (chauberahul1993) [reply svc-events/events/publishevttosubscriber.go:141]: this method get removed when reading from cache, Which is part of next PR

File: svc-events/events/getevtsubscription.go [hunk 16]
```diff
@@ -76,22 +76,22 @@ func (e *ExternalInterfaces) GetEventSubscriptionsDetails(req *eventsproto.Event
 		commonResponse := response.Response{
 			OdataType:    common.EventDestinationType,
 			ID:           evtSubscription.SubscriptionID,
-			Name:         evtSubscription.Name,
+			Name:         evtSubscription.EventDestination.Name,
```
[the remark this answers is not in the document]
Comment (chauberahul1993) [reply svc-events/events/getevtsubscription.go:79]: done

[the remark this answers is not in the document]
Comment (chauberahul1993) [reply svc-events/events/getevtsubscription.go:79]: done

File: svc-events/events/subscriptionCache.go (hunk 11 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go:146]: done

File: svc-events/events/subscriptionCache.go (hunk 2 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

File: svc-events/events/subscriptionCache.go (hunk 3 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

File: svc-events/events/subscriptionCache.go (hunk 7 above)
Comment (chauberahul1993) [reply svc-events/events/subscriptionCache.go]: done

**Pull request merged**